File: app/src/main/java/com/thibaudperso/sonycamera/timelapse/ui/settings/SettingsFragment.java
package com.thibaudperso.sonycamera.timelapse.ui.settings;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.Editable;
import android.text.Html;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.EditText;
import android.widget.TextView;

import com.thibaudperso.sonycamera.R;
import com.thibaudperso.sonycamera.timelapse.Constants;
import com.thibaudperso.sonycamera.timelapse.model.IntervalometerSettings;
import com.thibaudperso.sonycamera.timelapse.service.IntervalometerService;
import com.thibaudperso.sonycamera.timelapse.ui.processing.ProcessingActivity;

import static com.thibaudperso.sonycamera.R.id.settings_initial_delay;
import static com.thibaudperso.sonycamera.R.id.settings_interval_time;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;


public class SettingsFragment extends Fragment {


    private SharedPreferences mPreferences;
    
    private EditText mInitialDelayEditText;
    private EditText mIntervalTimeEditText;
    private EditText mFramesCountEditText;
    private CheckBox mFramesCountUnlimitedCheckBox;
    private TextView mErrorTextView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_settings, container, false);

        ((TextView) rootView.findViewById(R.id.settings_message)).setText(Html.fromHtml
                (getString(R.string.settings_message)));

        mErrorTextView = (TextView) rootView.findViewById(R.id.settings_error);

        /*
         * Initial delay field
         */

        mInitialDelayEditText = (EditText) rootView.findViewById(settings_initial_delay);
        mInitialDelayEditText.setText(String.valueOf(
                mPreferences.getInt(Constants.PREFERENCES_INITIAL_DELAY,
                        Constants.DEFAULT_INITIAL_DELAY)));
        createEditTextChangedListener(mInitialDelayEditText, new TextChangedListener() {
            @Override
            public void onTextChanged(String text) {
                setPreference(Constants.PREFERENCES_INITIAL_DELAY, getInitialDelay());
            }
        });



        /*
         * Interval time field
         */

        mIntervalTimeEditText = (EditText) rootView.findViewById(settings_interval_time);
        mIntervalTimeEditText.setText(String.valueOf(
                mPreferences.getInt(Constants.PREFERENCES_INTERVAL_TIME,
                        Constants.DEFAULT_INTERVAL_TIME)));
        createEditTextChangedListener(mIntervalTimeEditText, new TextChangedListener() {
            @Override
            public void onTextChanged(String text) {
                setPreference(Constants.PREFERENCES_INTERVAL_TIME, getIntervalTime());
            }
        });


        /*
         * Frames count fields
         */

        int framesCount = mPreferences.getInt(Constants.PREFERENCES_FRAMES_COUNT,
                Constants.DEFAULT_FRAMES_COUNT);
        boolean framesCountUnlimited = framesCount == 0;

        mFramesCountEditText = (EditText) rootView.findViewById(R.id.settings_frames_count);
        mFramesCountEditText.setText(framesCountUnlimited ? "" : String.valueOf(framesCount));
        mFramesCountEditText.setEnabled(!framesCountUnlimited);
        mFramesCountEditText.setFocusable(!framesCountUnlimited);
        mFramesCountEditText.setFocusableInTouchMode(!framesCountUnlimited);
        createEditTextChangedListener(mFramesCountEditText, new TextChangedListener() {
            @Override
            public void onTextChanged(String text) {
                setPreference(Constants.PREFERENCES_FRAMES_COUNT, getFramesCount());
            }
        });

        mFramesCountUnlimitedCheckBox = (CheckBox) rootView.findViewById(R.id.settings_frames_count_unlimited);
        mFramesCountUnlimitedCheckBox.setChecked(framesCountUnlimited);
        mFramesCountUnlimitedCheckBox.setOnCheckedChangeListener(new OnCheckedChangeListener() {

            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {

                mFramesCountEditText.setEnabled(!isChecked);
                mFramesCountEditText.setFocusable(!isChecked);
                mFramesCountEditText.setFocusableInTouchMode(!isChecked);

                setPreference(Constants.PREFERENCES_FRAMES_COUNT, getFramesCount());

            }
        });


        /*
         * Navigation buttons
         */
        rootView.findViewById(R.id.settings_previous).setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        getActivity().onBackPressed();
                    }
                });

        rootView.findViewById(R.id.settings_next).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(checkFields()) {
                    startService();
                }
            }
        });

        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();
        mErrorTextView.setVisibility(View.GONE);
    }

    private boolean checkFields() {

        if(getInitialDelay() < 0) {
            mErrorTextView.setVisibility(View.VISIBLE);
            mErrorTextView.setText(R.string.settings_initial_delay_positive_integer_error);
            return false;
        }

        if(getIntervalTime() <= 0) {
            mErrorTextView.setVisibility(View.VISIBLE);
            mErrorTextView.setText(R.string.settings_interval_time_positive_integer_error);
            return false;
        }

        if(getFramesCount() < 0) {
            mErrorTextView.setVisibility(View.VISIBLE);
            mErrorTextView.setText(R.string.settings_frames_count_positive_integer_error);
            return false;
        }

        mErrorTextView.setVisibility(View.GONE);
        return true;
    }


    private int getInitialDelay() {
        try {
            return Integer.parseInt(mInitialDelayEditText.getText().toString());
        } catch (NumberFormatException ignored) {
            return Constants.DEFAULT_INITIAL_DELAY;
        }
    }

    private int getIntervalTime() {
        try {
            return Integer.parseInt(mIntervalTimeEditText.getText().toString());
        } catch (NumberFormatException ignored) {
            return Constants.DEFAULT_INTERVAL_TIME;
        }
    }

    private int getFramesCount() {
        try {
            if(mFramesCountUnlimitedCheckBox.isChecked())
                return 0;
            return Integer.parseInt(mFramesCountEditText.getText().toString());
        } catch (NumberFormatException ignored) {
            return Constants.DEFAULT_FRAMES_COUNT;
        }
    }


    private void startService() {

//        ActivityCompat.finishAffinity(getActivity());
        Intent intent = new Intent(getContext(), ProcessingActivity.class);
        startActivity(intent);

        IntervalometerSettings settings = new IntervalometerSettings();
        settings.initialDelay = getInitialDelay();
        settings.intervalTime = getIntervalTime();
        settings.framesCount = getFramesCount();

        Intent serviceIntent = new Intent(getActivity(), IntervalometerService.class);
        serviceIntent.setAction(IntervalometerService.ACTION_START);
        serviceIntent.putExtra(IntervalometerService.EXTRA_SETTINGS, settings);
        getActivity().startService(serviceIntent);

    }

    private void setPreference(String prefName, Object prefValue) {

        SharedPreferences.Editor editor = mPreferences.edit();
        if (prefValue instanceof Integer) {
            editor.putInt(prefName, (Integer) prefValue);
        } else if (prefValue instanceof Float) {
            editor.putFloat(prefName, (Float) prefValue);
        } else if (prefValue instanceof Boolean) {
            editor.putBoolean(prefName, (Boolean) prefValue);
        }
        editor.apply();
    }


    /*
     * Shortcut for afterTextChanged listener
     */

    interface TextChangedListener {
        void onTextChanged(String text);
    }

    private void createEditTextChangedListener(EditText editText, final TextChangedListener listener) {
        editText.addTextChangedListener(new TextWatcher() {

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                listener.onTextChanged(s.toString());
            }
        });
    }

}

File: app/src/main/java/com/thibaudperso/sonycamera/timelapse/ui/settings/SettingsFragment.kt
package com.thibaudperso.sonycamera.timelapse.ui.settings

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.core.content.edit
import androidx.core.text.parseAsHtml
import com.thibaudperso.sonycamera.R
import com.thibaudperso.sonycamera.timelapse.BaseFragment
import com.thibaudperso.sonycamera.timelapse.Constants
import com.thibaudperso.sonycamera.timelapse.model.IntervalometerSettings
import com.thibaudperso.sonycamera.timelapse.service.IntervalometerService
import com.thibaudperso.sonycamera.timelapse.ui.processing.ProcessingActivity
import kotlinx.android.synthetic.main.fragment_settings.*

class SettingsFragment : BaseFragment() {

    private lateinit var sharedPreferences: SharedPreferences

    private val initialDelay: Long
        get() = try {
            settings_initial_delay.text.toString().toLong()
        } catch (ignored: NumberFormatException) {
            Constants.DEFAULT_INITIAL_DELAY
        }

    private val intervalTime: Long
        get() = try {
            settings_interval_time.text.toString().toLong()
        } catch (ignored: NumberFormatException) {
            Constants.DEFAULT_INTERVAL_TIME
        }

    private val framesCount: Int
        get() = try {
            if (settings_frames_count_unlimited.isChecked) {
                0
            } else {
                Integer.parseInt(settings_frames_count.text.toString())
            }
        } catch (ignored: NumberFormatException) {
            Constants.DEFAULT_FRAMES_COUNT
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_settings, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        settings_message.text = getString(R.string.settings_message).parseAsHtml()

        // Initial delay field
        settings_initial_delay.setText(sharedPreferences.getLong(Constants.PREFERENCES_INITIAL_DELAY,
                Constants.DEFAULT_INITIAL_DELAY).toString())
        settings_initial_delay.addTextChangeListener { setPreference(Constants.PREFERENCES_INITIAL_DELAY, initialDelay) }

        // Interval time field
        settings_interval_time.setText(sharedPreferences.getLong(Constants.PREFERENCES_INTERVAL_TIME,
                Constants.DEFAULT_INTERVAL_TIME).toString())
        settings_interval_time.addTextChangeListener { setPreference(Constants.PREFERENCES_INTERVAL_TIME, intervalTime) }

        // Frames count fields
        val framesCount = sharedPreferences.getInt(Constants.PREFERENCES_FRAMES_COUNT, Constants.DEFAULT_FRAMES_COUNT)
        val framesCountUnlimited = framesCount == 0

        settings_frames_count.setText(if (framesCountUnlimited) "" else framesCount.toString())
        settings_frames_count.isEnabled = !framesCountUnlimited
        settings_frames_count.isFocusable = !framesCountUnlimited
        settings_frames_count.isFocusableInTouchMode = !framesCountUnlimited
        settings_frames_count.addTextChangeListener { setPreference(Constants.PREFERENCES_FRAMES_COUNT, framesCount) }

        settings_frames_count_unlimited.isChecked = framesCountUnlimited
        settings_frames_count_unlimited.setOnCheckedChangeListener { _, isChecked ->
            settings_frames_count.isEnabled = !isChecked
            settings_frames_count.isFocusable = !isChecked
            settings_frames_count.isFocusableInTouchMode = !isChecked
            setPreference(Constants.PREFERENCES_FRAMES_COUNT, framesCount)
        }

        // Navigation buttons
        settings_previous.setOnClickListener { requireActivity().onBackPressed() }
        settings_next.setOnClickListener {
            if (checkFields()) {
                startService()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        settings_error.visibility = View.GONE
    }

    private fun checkFields(): Boolean {
        if (initialDelay < 0) {
            settings_error.visibility = View.VISIBLE
            settings_error.setText(R.string.settings_initial_delay_positive_integer_error)
            return false
        }

        if (intervalTime <= 0) {
            settings_error.visibility = View.VISIBLE
            settings_error.setText(R.string.settings_interval_time_positive_integer_error)
            return false
        }

        if (framesCount < 0) {
            settings_error.visibility = View.VISIBLE
            settings_error.setText(R.string.settings_frames_count_positive_integer_error)
            return false
        }

        settings_error.visibility = View.GONE
        return true
    }

    private fun startService() {
        val intent = Intent(context, ProcessingActivity::class.java)
        startActivity(intent)

        val settings = IntervalometerSettings()
        settings.initialDelay = this.initialDelay
        settings.intervalTime = this.intervalTime
        settings.framesCount = framesCount

        val serviceIntent = Intent(activity, IntervalometerService::class.java)
        serviceIntent.action = IntervalometerService.ACTION_START
        serviceIntent.putExtra(IntervalometerService.EXTRA_SETTINGS, settings)
        requireActivity().startService(serviceIntent)
    }

    private fun setPreference(prefName: String, prefValue: Any) {
        sharedPreferences.edit {
            when (prefValue) {
                is Int -> putInt(prefName, prefValue)
                is Float -> putFloat(prefName, prefValue)
                is Boolean -> putBoolean(prefName, prefValue)
            }
        }
    }

    private fun EditText.addTextChangeListener(listener: (Any) -> Unit) = addTextChangedListener(object : TextWatcher {
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) = Unit
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit
        override fun afterTextChanged(s: Editable) {
            listener(s.toString())
        }
    })

}
